// @superset-ui/legacy-plugin-chart-markup-data/esm/Markup.js
"use strict";

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const propTypes = {
  className: _propTypes.default.string,
  cssFiles: _propTypes.default.arrayOf(_propTypes.default.string),
  height: _propTypes.default.number.isRequired,
  html: _propTypes.default.string,
  isSeparator: _propTypes.default.bool
};
const defaultProps = {
  className: '',
  cssFiles: [],
  html: '',
  isSeparator: false
};
const CONTAINER_STYLE = {
  overflow: 'auto',
  position: 'relative'
};

class Markup extends _react.default.PureComponent {
  render() {
    const {
      className,
      height,
      isSeparator,
      html,
      cssFiles
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: className,
      style: CONTAINER_STYLE
    }, /*#__PURE__*/_react.default.createElement("iframe", {
      title: "superset-markup",
      frameBorder: 0,
      height: isSeparator ? height - 20 : height,
      sandbox: "allow-forms allow-popups allow-same-origin allow-scripts allow-top-navigation",
      srcDoc: "\n            <html>\n              <head>\n                " + cssFiles.map(href => "<link rel=\"stylesheet\" type=\"text/css\" href=\"" + href + "\" />") + "\n              </head>\n              <body style=\"background-color: transparent;\">\n                " + html + "\n              </body>\n            </html>"
    }));
  }

}

Markup.propTypes = propTypes;
Markup.defaultProps = defaultProps;

export default Markup;
